catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserverTest.java

```diff
@@ -0,0 +1,1007 @@
+
+    runThreads(observer::checkAndNotify, 10);
+
+    verify(fileListener, times(0)).onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length))
+        .onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(0)).onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testDeleteThreadSafety() throws Exception {
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+    observer.checkAndNotify();
+    init();
+
+    Stream.of(files).forEach(this::fileDelete);
+
+    runThreads(observer::checkAndNotify, 10);
+
+    verify(fileListener, times(0)).onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(0)).onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length))
+        .onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testModifyAndDeleteWithDelays() throws Exception {
+    int delay = 200;
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+    observer.checkAndNotify();
+    init();
+
+    Stream.of(files).forEach(this::changeData);
+
+    addDelay(delay);
+
+    observer.checkAndNotify();
+
+    Stream.of(files).forEach(this::fileDelete);
+    //  Implementation Detail
+    //  Because the last operation hasn't finished yet the file should not be deleted in the state.
+    //  It will wait until the file has finished it's prior update to remove it.
+
+    observer.checkAndNotify();
+    verify(fileListener, times(0)).onFileDelete(any(File.class), any(Synchronization.class));
+    Thread.sleep(delay + delayBuf);
+
+    verify(fileListener, times(files.length))
+        .onFileChange(any(File.class), any(Synchronization.class));
+
+    removeDelay();
+
+    observer.checkAndNotify();
+
+    verify(fileListener, times(0)).onFileCreate(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length))
+        .onFileChange(any(File.class), any(Synchronization.class));
+    verify(fileListener, times(files.length))
+        .onFileDelete(any(File.class), any(Synchronization.class));
+  }
+
+  @Test
+  public void testCreateAndDeleteWithDelays() throws Exception {
+    int delay = 200;
+
+    File[] files = initFiles(1, monitoredDirectory, "file00");
+
+    addDelay(delay);
+
+    observer.checkAndNotify();
+
+    Stream.of(files).forEach(this::fileDelete);
+    //  Because the last operation hasn't finished yet the file should not be deleted in the state.
+    //  It will wait until the file has finished it's prior update to remove it.
+
+    observer.checkAndNotify();
+    verify(fileListener, times(0)).onFileDelete(any(File.class), any(Synchronization.class));
+
+    Thread.sleep(delay + delayBuf);
+
+    verify(fileListener, times(files.length))
+        .onFileCreate(any(File.class), any(Synchronization.class));
+
+    removeDelay();
```

Comment:
✏️ Does the delay need to be removed at the end of every test after `addDelay`? `doTestWrapper` will be rest after every `@Test` method.

Reply:
The original idea was if there was something that was still waiting for some reason then do one last pass without a delay to catch it.